File: services/quiz-engine/src/answer_validator.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any


class AnswerValidator(ABC):
    """Protocol for validating answers based on question type."""

    @abstractmethod
    def is_correct(self, correct_data: Any, selected: Any) -> bool:
        """Return True if the answer is correct."""
        ...


class MCQValidator(AnswerValidator):
    """Validates MCQ answers: all correct choices must be selected, no extras."""

    def is_correct(self, correct_data: list[int], selected: list[int]) -> bool:
        if not isinstance(correct_data, list) or not isinstance(selected, list):
            return False
        return set(selected) == set(correct_data)


# Registry for future question types
VALIDATORS: dict[str, AnswerValidator] = {
    "MCQ": MCQValidator(),
}


def get_validator(question_type: str) -> AnswerValidator:
    validator = VALIDATORS.get(question_type)
    if validator is None:
        raise ValueError(f"Unknown question type: {question_type}")
    return validator

File: services/web-backend/src/engine/answer-validator.ts
/* ------------------------------------------------------------------ */
/*  Answer validators — Strategy pattern for question types            */
/*  Port of answer_validator.py                                        */
/* ------------------------------------------------------------------ */

export interface AnswerValidator {
  /** Return true when the answer is considered correct. */
  isCorrect(correctData: unknown, selected: unknown): boolean
}

/**
 * Validates MCQ (multiple-choice) answers:
 * - All correct choices must be selected.
 * - No extra (wrong) choices may be selected.
 */
export class MCQValidator implements AnswerValidator {
  isCorrect(correctData: unknown, selected: unknown): boolean {
    if (!Array.isArray(correctData) || !Array.isArray(selected)) {
      return false
    }
    if (correctData.length === 0) {
      return selected.length === 0
    }
    const correctSet = new Set(correctData)
    const selectedSet = new Set(selected)
    if (correctSet.size !== selectedSet.size) return false
    for (const c of correctSet) {
      if (!selectedSet.has(c)) return false
    }
    return true
  }
}

// ── Registry ──────────────────────────────────────────────────────────

export const VALIDATORS: Record<string, AnswerValidator> = {
  MCQ: new MCQValidator(),
}

export function getValidator(questionType: string): AnswerValidator {
  const validator = VALIDATORS[questionType]
  if (!validator) {
    throw new Error(`Unknown question type: ${questionType}`)
  }
  return validator
}

/**
 * Convenience wrapper: checks whether the selected choices are correct
 * for the given question type.
 */
export function validateAnswer(
  questionType: string,
  correctChoices: number[],
  selectedChoices: number[],
): boolean {
  return getValidator(questionType).isCorrect(correctChoices, selectedChoices)
}
